Potential enhancement to support overrides on translated attributes in derived classes

deltafsdevelopment opened the pull request:
First off - many thanks @damieng for creating this excellent extension to Linq-to-entities I have been using it for the passed 5 years with great success. 

The only niggle has been around mapping calculated attributes in derived classes where I want to override the attribute mapped in the base class with a new calculation in the derived class. This didn't work out of the box.

This set of changes represent my fix/enhancement for that functionality.

I have included an nUnit test project to demonstrate the changes which utilises the excellent Effort in memory database stub.

Unfortunately I am still on VS 2013 so in order to demo these changes I had to unwind the upgrade to VS 2015 and project.json ...

damieng commented:
Hi delta.  Interesting additional although I'm not keen on reverting entirely back to VS2013. Might be worth having a dual-project system to support this.

deltafsdevelopment commented:
Sure understand - really this fork is just to demo the change - the adjustments are actually very minor, in `TranslationMap.Add` I have changed the logic to store the derived class name:

```
        public void Add<T, TResult>(Expression<Func<T, TResult>> property, CompiledExpression<T, TResult> compiledExpression)
        {
            Argument.EnsureNotNull("property", property);
            Argument.EnsureNotNull("compiledExpression", compiledExpression);
            
            // deltafsdevelopment May 2013 - Fix to the code here so that if the prop is an override the derived class type name is
            // is stored in the translation map not the base class name that way we can have different expressions mapped to the
            // same override in different derived classes
            if (property.Parameters.Count > 0)
            {
              ParameterExpression expr = property.Parameters[0];
              PropertyInfo pi = expr.Type.GetProperty(((MemberExpression)property.Body).Member.Name);
              base.Add(pi, compiledExpression);
            }
        }
```

And in `TranslatingVisitor.VisitMember` I have added logic to _walk back up the inheritance hierarchy_ to find the translation:

```
            /// <summary>
            ///  Walk up the inheritance hierarchy searching for a compiled expression attached to a
            ///  property of the given name
            /// </summary>
            /// <param name="propName">Name of the property to search for</param>
            /// <param name="type">Type of the member to search against</param>
            /// <returns>Compiled expression if found or null if not</returns>
            private CompiledExpression FindCompiledExpression(String propName, Type type)
            {
                while (type != typeof(Object))
                {
                    CompiledExpression cp;
                    MemberInfo mi = type.GetProperty(propName, BindingFlags.DeclaredOnly | BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);
                    EnsureTypeInitialized(type);
                    if (mi != null && map.TryGetValue(mi, out cp))
                    {
                        return cp;
                    }
                    type = type.BaseType;
                }
                return null;
            }

            protected override Expression VisitMember(MemberExpression node)
            {
                Argument.EnsureNotNull("node", node);
                // deltafsdevelopment 10 Nov 2015 - Fix to original code here so that the code searches for CompiledExpressions
                // right through the inheritance hierarchy to allow them to be defined on base classes and on overrides
                if (node.Expression != null)
                {
                    Type type = node.Expression.Type;
                    String propName = node.Member.Name;
                    CompiledExpression cp = FindCompiledExpression(propName, type);
                    if (cp != null)
                    {
                        return VisitCompiledExpression(cp, node.Expression);
                    }
                }
                return base.VisitMember(node);

            }
```

As I have been using the modification trouble free for a couple of years now I thought it was only courteous to share it.

deltafsdevelopment commented:
When my team (eventually) upgrade to VS 2015 I will resubmit the pull thanks
